### DISCUSSION TITLE
got Future <Future pending> attached to a different loop

### DISCUSSION THREAD

**Discussion by krishsharma0413** [fastapi/fastapi#6566] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
def main():
  uvicorn.run(app,host="0.0.0.0",port=8090,loop="asyncio")
loop = asyncio.get_event_loop()
def run():
  loop.run_until_complete(main())

def keep_alive1():
    t = threading.Thread(target=run)
    t.start()

keep_alive1()


client.run('discord bot token')
```

### Description

i was using discord client with fastapi i used another thread for running uvicorn.
the bot worked fine but when i used the fastapi endpoints it gave me an error
```
RuntimeError: Task <Task pending name='Task-32' coro=<RequestResponseCycle.run_asgi() running at /opt/virtualenvs/python3/lib/python3.8/site-packages/uvicorn/protocols/http/h11_impl.py:369> cb=[set.discard()]> got Future <Future pending> attached to a different loop
```
the endpoint that gave me this error was asynchronous using mongodb module motor, other non asynchronous  endpoints worked fine like the /docs.

when i removed the discord client with just the fastapi it still didnt worked with uvicorn.run() but did worked with the cmd command uvicorn main:app not sure why tho.

### Operating System

Windows

### FastAPI Version

0.66.0

### Python Version

3.9.4

**Comment by frankie567** [comment, marked as answer]
This is a quite common issue when using Motor (unfortunately, their implementation doesn't always play well with some event loops patterns). I wrote some details about this some time ago here: https://github.com/fastapi-users/fastapi-users/discussions/663#discussioncomment-1025558

When using Motor, if you want to invoke Uvicorn programmatically you have to make sure of two things:
1. Put the invokation in a separate file (like `main.py`) alone, not besides the `app` instance;
2. Don't import `app` to pass it to uvicorn: prefer the import path as a string. This way, you'll ensure Uvicorn imports the module by itself with the right loop attached:

```py
# main.py

import uvicorn

if __name__ == "__main__":
    # Assuming your `app` variable is in a module named `app.py`
    uvicorn.run("app:app", host="0.0.0.0", port=8000)
```

BTW, I see in your example that you do `loop.run_until_complete(main())` on the function that calls Uvicorn. You don't need to do this (see above), Uvicorn takes care of this.

**Comment by Kludex** [comment]
Is this related to https://github.com/encode/uvicorn/pull/1110 ?

**Comment by frankie567** [comment]
From what I understand of the issue, yes, this is what probably causes the problem.

However, I also think that Motor could greatly improve things on their side: contrary to most database drivers, they don't provide a connect/disconnect mechanism we could trigger on FastAPI startup; which would be then in the right loop context.

**Comment by Kludex** [comment]
Yeah, that PR solves an issue that should never happen. :)

**Comment by krishsharma0413** [comment]
thanks it worked, tho you should add a little note in the docs about the motor connector.

**Comment by rbrisita** [comment]
The issue could also be creating `asyncio.Queue()` outside of the current event loop. These queues should be created inside the main coroutine so they access the current running loop.

**Reply by kevinddchen** [reply]
Thank you. I got an error similar to this one, and not creating a queue until inside the event loop fixed the root issue.
